fact_extractor/plugins/unpacking/generic_carver/code/generic_carver.py

```diff
@@ -5,145 +5,100 @@
-
-
-def _output_is_empty(output):
-    return int((output.split())[-1]) == 0
-
-
-def _find_trailing_data_index_zip(file_path: Path) -> int | None:
-    """Archives carved by binwalk often have trailing data at the end. 7z can determine the actual file size."""
-    output = execute_shell_command(f'7z l {file_path}')
-    if 'There are data after the end of archive' in output:
-        match = REAL_SIZE_REGEX.search(output)
-        if match:
-            return int(match.groups()[0])
-    return None
-
-
-def _find_trailing_data_index_bz2(file_path: Path) -> int | None:
-    output = execute_shell_command(f'bzip2 -t {file_path}')
-    if 'trailing garbage' in output:
-        file_content = file_path.read_bytes()
-        matches = sorted(index for magic in BZ2_EOF_MAGIC if (index := file_content.find(magic)) != -1)
-        # there may be two matches, but we want the first one (but also not -1 == no match)
-        if matches:
-            # 10 is magic string + CRC 32 checksum + padding (see https://en.wikipedia.org/wiki/Bzip2#File_format)
-            return matches[0] + 10
-    return None
-
-
-def drop_underscore_directory(tmp_dir):
-    extracted_contents = list(Path(tmp_dir).iterdir())
-    if not extracted_contents:
-        return
-    if len(extracted_contents) != 1 or not extracted_contents[0].name.endswith('.extracted'):
-        return
-    for result in extracted_contents[0].iterdir():
-        shutil.move(str(result), str(result.parent.parent))
-    shutil.rmtree(str(extracted_contents[0]))
+VERSION = '1.0.0'
+
+MIN_FILE_ENTROPY = 0.01
+
+# deactivate internal logger of unblob because it can slow down searching chunks
+structlog.configure(wrapper_class=structlog.make_filtering_bound_logger(logging.CRITICAL))
+
+
+# register custom zlib handler to allow carving zlib chunks from inside files
+@hookimpl
+def unblob_register_handlers():
+    yield from [ZlibCarvingHandler]
+
+
+class ZlibCarvingHandler(ZlibHandler):
+    NAME = 'zlib_carver'
+
+    PATTERNS = [  # noqa: RUF012
+        HexString('78 01'),  # low compression
+        HexString('78 9c'),  # default compression
+        HexString('78 da'),  # best compression
+        HexString('78 5e'),  # compressed
+    ]
+
+
+def unpack_function(file_path: str, tmp_dir: str) -> dict:
+    extraction_dir = Path(tmp_dir)
+    chunks = []
+    filter_report = ''
+    path = Path(file_path)
+
+    try:
+        with File.from_path(path) as file:
+            for chunk in _find_chunks(path, file):
+                if isinstance(chunk, PaddingChunk):
+                    continue
+                if isinstance(chunk, UnknownChunk):
+                    if _has_low_entropy(file, chunk):
+                        filter_report += (
+                            f'removed chunk {chunk.start_offset}-{chunk.end_offset} (reason: low entropy)\n'
+                        )
+                        continue
+                    carve_unknown_chunk(extraction_dir, file, chunk)
+                else:
```

Comment:
I take it there are only three chunk types (`PaddingChunk`, `UnknownChunk` and the implicit valid Chunk)?

Reply:
It seems like it (at least according to `unblob/models.py`)